test/envtest/konglicense_controller_test.go

```diff
@@ -0,0 +1,126 @@
+package envtest
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/samber/lo"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/controllers/configuration"
+	kongv1alpha1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1alpha1"
+)
+
+func TestKongLicenseController(t *testing.T) {
+	scheme := Scheme(t, WithKong)
+	cfg := Setup(t, scheme)
+	ctrlClient := NewControllerClient(t, scheme, cfg)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	reconciler := &configuration.KongV1Alpha1KongLicenseReconciler{
+		Client:         ctrlClient,
+		Log:            logr.Discard(),
+		Scheme:         scheme,
+		LicenseCache:   configuration.NewLicenseCache(),
+		ControllerName: "test",
+	}
+	StartReconcilers(ctx, t, ctrlClient.Scheme(), cfg, reconciler)
+
+	const (
+		waitTime            = 3 * time.Second
+		tickTime            = 100 * time.Millisecond
+		fullControllerName  = configuration.LicenseControllerType + "/test"
+		conditionProgrammed = "Programmed"
+	)
+
+	t.Log("Create a KongLicense and verify that it is reconciled")
+	kongLicense1 := &kongv1alpha1.KongLicense{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "license-1",
+		},
+		RawLicenseString: "test-license-1",
+		Enabled:          true,
+	}
+	err := ctrlClient.Create(ctx, kongLicense1)
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		l, ok := reconciler.GetLicense().Get()
+		if !ok {
+			return false
+		}
+		return l.Payload != nil && *l.Payload == "test-license-1"
+	}, waitTime, tickTime, "Should return the license in GetLicense method")
+
+	require.Eventually(t, func() bool {
+		err := ctrlClient.Get(ctx, k8stypes.NamespacedName{Name: kongLicense1.Name}, kongLicense1)
+		require.NoError(t, err, "Should not have error in getting the latest status")
+		return findConditionInControllerStatus(
+			kongLicense1, fullControllerName, conditionProgrammed, metav1.ConditionTrue)
+	}, waitTime, tickTime, "The Programmed condition in controller status should be true")
+
+	t.Log("Wait for a second and create a new KongLicense, then verify it replaces the old one")
+	time.Sleep(time.Second)
```

Comment:
Why do we need to wait here?

Reply:
Because if we do not sleep, the `creationTimestamp` may be the same. The `creationTimestamp` is in precision of seconds.

Reply:
Sorry for not having the context here but I believe this would be worth documenting. It's not immediately clear to the reader why the "replacement" happens. I assume it's because the reconciler takes the license that's the most recent but that would be worth adding as a log in the test.

TIL: there was a proposal to change the resolution but got nowhere apparently: https://github.com/kubernetes/kubernetes/issues/81026

Reply:
> The creationTimestamp is in precision of seconds.
TIL. Wasn't aware of that. 

@pmalek I'm not sure if Tao will be able to address this now as he's already gone on PTO. I've added a comment with some rationale behind the test steps.